### PR TITLE
[ctx_prof] Flattened profile lowering pass

### PR DESCRIPTION
**Pull request by mtrofin**
Pass to flatten and lower the contextual profile to profile (i.e. `MD_prof`) metadata. This is expected to be used after all IPO transformations have happened.

Prior to lowering, the instrumentation is maintained during IPO and the contextual profile is kept in sync (see PRs #105469, #106154). Flattening (#104539) sums up all the counters belonging to all a function's context nodes.

We first propagate counter values (from the flattened profile) using the same propagation algorithm as `PGOUseFunc::populateCounters`, then map the edge values to `branch_weights`. Functions. in the module that don't have an entry in the flattened profile are deemed cold, and any `MD_prof` metadata they may have is reset. The profile summary is also reset at this point.

Issue [#89287](https://github.com/llvm/llvm-project/issues/89287)

### CODE REVIEW

**Comment by mtrofin** [comment]
* **#107329** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/107329?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @mtrofin and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: llvm/test/Analysis/CtxProfAnalysis/flatten-and-annotate.ll [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+;  - no instrumentation
+;  - the 2 functions have an entry count
+;  - each conditional branch has profile annotation
+;
+; CHECK-NOT:   call void @llvm.instrprof
+;
+; make sure we have function entry counts, branch weights, and a profile summary.
+; CHECK-LABEL: @foo
+; CHECK-SAME:    !prof !29
+; CHECK:          br i1 %t, label %yes, label %no, !prof !31
```
Comment (david-xl) [inline llvm/test/Analysis/CtxProfAnalysis/flatten-and-annotate.ll:34]: Use name references instead of hard coding !31.

File: llvm/test/Analysis/CtxProfAnalysis/flatten-and-annotate.ll [hunk 2]
```diff
@@ -0,0 +1,88 @@
+; REQUIRES: x86_64-linux
```
Comment (david-xl) [inline llvm/test/Analysis/CtxProfAnalysis/flatten-and-annotate.ll:1]: Can the test case be augmented to show counter merging (from different contexts) after flattening?

Comment (mtrofin) [reply llvm/test/Analysis/CtxProfAnalysis/flatten-and-annotate.ll:1]: ptal

File: llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+    return PreservedAnalyses::all();
+
+  const auto FlattenedProfile = CtxProf.flatten();
+
+  InstrProfSummaryBuilder PB(ProfileSummaryBuilder::DefaultCutoffs);
+  for (auto &F : M) {
+    if (F.isDeclaration())
+      continue;
+
+    if (!areAllBBsReachable(F,
```
Comment (david-xl) [inline llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp:307]: Why skipping those functions? Can they be handled differently?

File: llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+            "functions (message pumps, usually) which are not supported in the "
+            "contextual profiling case");
+      setProfMetadata(F.getParent(), Term, EdgeCounts, MaxCount);
+    }
+  }
+};
+
+bool areAllBBsReachable(const Function &F, FunctionAnalysisManager &FAM) {
+  auto &DT = FAM.getResult<DominatorTreeAnalysis>(const_cast<Function &>(F));
+  for (const auto &BB : F)
```
Comment (david-xl) [inline llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp:273]: use llvm::any_of ?

File: llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+    if (!DT.isReachableFromEntry(&BB))
+      return false;
+  return true;
+}
+
+void clearColdFunctionProfile(Function &F) {
+  for (auto &BB : F)
```
Comment (david-xl) [inline llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp:280]: is this step needed? Setting entry point count seems enough.

File: llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+    BB.getTerminator()->setMetadata(LLVMContext::MD_prof, nullptr);
+  F.setEntryCount(0U);
+}
+
+void removeInstrumentation(Function &F) {
```
Comment (david-xl) [inline llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp:285]: why is this needed in profile-use compilation?

File: llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp (hunk 3 above)
Comment (mtrofin) [reply llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp:307]: we don't skip, we error. The expectation is that we'd run a DCE pass before. It looks like we skip because the control flow needs to continue, but `emitError` should result in an `exit`.

File: llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp (hunk 4 above)
Comment (mtrofin) [reply llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp:273]: good point. or `all_of`.

File: llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp (hunk 5 above)
Comment (mtrofin) [reply llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp:280]: for consistency.

File: llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp (hunk 6 above)
Comment (mtrofin) [reply llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp:285]: We instrument in both cases. In the "profile use" case, the instrumentation serves as anchors to which we match counters - makes maintaining the association straight forward during IPO.

File: llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+    SmallVector<EdgeInfo *> OutEdges;
+    SmallVector<EdgeInfo *> InEdges;
+    size_t UnknownCountOutEdges = 0;
+    size_t UnknownCountInEdges = 0;
+
+    uint64_t getEdgeSum(const SmallVector<EdgeInfo *> &Edges,
+                        bool AssumeAllKnown) const {
+      uint64_t Sum = 0;
+      for (const auto *E : Edges)
+        if (E)
```
Comment (david-xl) [inline llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp:65]: Why can E be null?

File: llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp [hunk 8]
```diff
@@ -0,0 +1,341 @@
...
+  };
+
+  class BBInfo {
+    std::optional<uint64_t> Count;
+    SmallVector<EdgeInfo *> OutEdges;
+    SmallVector<EdgeInfo *> InEdges;
+    size_t UnknownCountOutEdges = 0;
+    size_t UnknownCountInEdges = 0;
+
+    uint64_t getEdgeSum(const SmallVector<EdgeInfo *> &Edges,
```
Comment (david-xl) [inline llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp:61]: brief document when to assumeAllKnown.

File: llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp [hunk 9]
```diff
@@ -0,0 +1,341 @@
...
+                        bool AssumeAllKnown) const {
+      uint64_t Sum = 0;
+      for (const auto *E : Edges)
+        if (E)
+          Sum += AssumeAllKnown ? *E->Count : E->Count.value_or(0U);
+      return Sum;
+    }
+
+    void takeCountFrom(const SmallVector<EdgeInfo *> &Edges) {
```
Comment (david-xl) [inline llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp:70]: nit: take-> compute

File: llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp (hunk 6 above)
Comment (david-xl) [reply llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp:284]: right -- I saw the comment at the top.

File: llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp (hunk 3 above)
Comment (david-xl) [reply llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp:316]: Error is intended for users which they not do anything about it. Perhaps just assert?

Comment (mtrofin) [reply llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp:316]: I see - yes.

File: llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp (hunk 5 above)
Comment (mtrofin) [reply llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp:281]: This is only run in the module(s) that the profile is applicable to, though. So the expectation is that most function are participating in the profile (because ThinLTO put them there).

I'd prefer keeping things consistent here and revisiting if, indeed, compile time due to this is an issue. The value of consistency (here) is that one doesn't need to know to check first the function entry count to make sense of the branch weights; also, by eliding any weights - should there be any - we help ulterior BFI updates (ok, unless those also discard any entrycount == 0 functions)

File: llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp (hunk 8 above)
Comment (mtrofin) [reply llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp:63]: done.

File: llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp (hunk 7 above)
Comment (mtrofin) [reply llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp:67]: We place edges at their position in the terminator list - i.e. in the position that edge's count must appear in the `branch_weights` vector. Some edges - currently, the coro suspend ones - are not real. PGO marks those as removed during MST buildup, but we don't do that here, so using `nullptr` instead.

Comment (mtrofin) [reply llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp:67]: To be clear, this is applicable to the `Out` edges, for the `In` edges we don't care about position.

**Review by david-xl — APPROVED** [review]

**Comment by mtrofin** [comment]
### Merge activity

* **Sep 6, 4:40 PM EDT**: @mtrofin started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/107329).
